# <font color='#d35400'> Lab 5 | Hypothesis Testing </font>
Welcome to Lab 5! In this lab we work on hypothesis testing by implementing several different hypothesis tests from scratch and then comparing the results we get to the outputs of python functions. This lab explores using the World Health Organization (WHO) data which tracks a number of useful health-related metrics aggregated at the country level.

<p align="center">
  <img src="dog_loves_sausages.jpeg" alt="Alt Text", width="300"/>
</p>

### <font color='#FF8C00'> About the Dataset </font>
We start off this Jupyter Notebook by first examining the dataset from `who2009.csv`. We look at the features that are in the dataset, as well as the data types being used in each of the features. We would like to get an idea of what we're dealing with, so we start off by displaying the dataset in Pandas.

In [1]:
# importing the pandas library
import pandas as pd

# importing the ipython library
from IPython.display import display

# read the .csv file using the pandas library
who_df = pd.read_csv("who2009.csv")

# displaying the pandas dataframe
who_df

country  v1  v2  v3  v4  v5  v6  v7  v8  v9  \
0                           Afghanistan  41  40  41  42  42  42  42  41  42   
1                               Albania  66  68  71  71  73  74  69  71  72   
2                               Algeria  65  68  70  68  71  73  66  70  71   
3                               Andorra  74  76  78  81  83  85  77  80  81   
4                                Angola  38  46  51  44  51  55  41  49  53   
..                                  ...  ..  ..  ..  ..  ..  ..  ..  ..  ..   
188  Venezuela (Bolivarian Republic of)  70  71  72  74  77  78  72  74  75   
189                            Viet Nam  64  68  70  68  72  75  66  70  72   
190                               Yemen  56  59  62  59  63  66  58  61  64   
191                              Zambia  52  41  45  55  44  47  53  42  46   
192                            Zimbabwe  57  43  45  63  46  44  60  44  45   

     ...    v257    v258     v259  v260  countrycode  iso3  regioncode  \
0    ...     NaN     NaN      NaN   NaN            4   AFG         142   
1    ...  2540.0  3920.0   6580.0   1.0            8   ALB         150   
2    ...  4350.0  5130.0   7640.0   NaN           12   DZA           2   
3    ...     NaN     NaN      NaN   NaN           20   AND         150   
4    ...  1870.0  1910.0   4400.0  42.5           24   AGO           2   
..   ...     ...     ...      ...   ...          ...   ...         ...   
188  ...  6820.0  8380.0  11920.0  10.0          862   VEN          19   
189  ...   610.0  1400.0   2550.0  22.8          704   VNM         142   
190  ...  1280.0  1710.0   2200.0  17.5          887   YEM         142   
191  ...   820.0   870.0   1220.0  64.3          894   ZMB           2   
192  ...     NaN     NaN      NaN   NaN          716   ZWE           2   

     regionname  subregioncode       subregionname  
0          Asia             34       Southern Asia  
1        Europe             39     Southern Europe  
2        Africa             15     Northern Africa  
3        Europe             39     Southern Europe  
4        Africa             17       Middle Africa  
..          ...            ...                 ...  
188    Americas              5       South America  
189        Asia             35  South-Eastern Asia  
190        Asia            145        Western Asia  
191      Africa             14      Eastern Africa  
192      Africa             14      Eastern Africa  

[193 rows x 267 columns]

In [2]:
# displaying the descriptive statistics of the dataset
display("Summary Statistics ", who_df.info())

# displaying the summary of the dataset
display("Description of Dataset", who_df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Columns: 267 entries, country to subregionname
dtypes: float64(209), int64(53), object(5)
memory usage: 402.7+ KB


'Summary Statistics '

None

'Description of Dataset'

v1          v2          v3          v4          v5          v6  \
count  193.000000  193.000000  193.000000  193.000000  193.000000  193.000000   
mean    62.145078   63.860104   65.502591   67.310881   68.875648   70.274611   
std      9.752654    9.834363    9.813428   10.199404   10.690975   10.754952   
min     25.000000   34.000000   39.000000   37.000000   42.000000   42.000000   
25%     57.000000   57.000000   58.000000   61.000000   62.000000   63.000000   
50%     64.000000   66.000000   67.000000   70.000000   73.000000   74.000000   
75%     69.000000   71.000000   72.000000   75.000000   77.000000   78.000000   
max     76.000000   78.000000   81.000000   82.000000   85.000000   86.000000   

               v7          v8          v9         v10  ...        v254  \
count  193.000000  193.000000  193.000000  193.000000  ...  163.000000   
mean    64.626943   66.367876   67.839378   58.595855  ...   87.128834   
std      9.956975   10.207963   10.223548    9.363181  ...   13.090920   
min     31.000000   38.000000   41.000000   34.000000  ...   32.000000   
25%     58.000000   60.000000   61.000000   52.000000  ...   83.500000   
50%     67.000000   70.000000   71.000000   60.000000  ...   91.000000   
75%     72.000000   74.000000   75.000000   65.000000  ...   96.000000   
max     79.000000   81.000000   83.000000   74.000000  ...  100.000000   

             v255        v256          v257          v258          v259  \
count  124.000000  164.000000    153.000000    172.000000    175.000000   
mean    81.491935   84.902439   5999.869281   8424.825581  12026.742857   
std     20.795362   17.016852   7204.320249  10187.348849  13386.001361   
min     21.000000    0.000000    250.000000    210.000000    290.000000   
25%     77.000000   82.000000   1190.000000   1427.500000   2220.000000   
50%     90.500000   90.500000   2970.000000   4385.000000   6580.000000   
75%     96.000000   96.000000   8100.000000   9860.000000  16515.000000   
max    100.000000  100.000000  38760.000000  46510.000000  63590.000000   

            v260  countrycode  regioncode  subregioncode  
count  93.000000   193.000000  193.000000     193.000000  
mean   25.901075   431.181347   71.963731      83.777202  
std    25.445049   254.403312   68.744195     106.063990  
min     0.000000     4.000000    2.000000       5.000000  
25%     1.200000   208.000000    2.000000      14.000000  
50%    19.700000   428.000000   19.000000      35.000000  
75%    46.100000   646.000000  142.000000     145.000000  
max    86.100000   894.000000  150.000000     419.000000  

[8 rows x 262 columns]

### <font color='#FF8C00'> Dataset Information & Summary </font>
According `.info()` and `.describe()`, we can see that there are 193 rows & 267 columns. With regards to te column headings on the dataset, they are presented as follows:
- **Country:** Member State of the World Health Organization
- **v9:** life expectancy at birth for both sexes
- **v22:** infant mortality rate of both sexes
- **v159:** health workforce
- **v168:** hospital beds
- **v174:** total expenditure on health as % of GDP
- **v186:** out of pocket expenditure as % of private expenditure on health
- **v192:** per capita total expenditure on health
- **v249:** total fertility rate
- **v259:** gross national income per capita 
- **regionname:** alphabetic ragion name

## <font color = '#FF8C00'> Section 1 </font> | Data Preparation
In this section, we work on reading a `.csv` file in Pandas, and then perform some analysis, such as looking at the size of the dataset and understand what the features are by referring to the code book. We then rename the columns according to the code book. The instructions are arranged in the bullet points as follows:
- [x] First read the who2009.csv file into your python development environment using pandas. You’ll need to import pandas and then apply the `.read csv` method.
- [x] Assess the size of the data set using the .shape method. You should notice that there are a large number of columns in the data. We will only need a very small number of them. I have provided a codebook to go with this data (WHO2009SubsetCodebook.pdf). Drop columns from the data, keeping only those identified in the codebook.
- [ ] Rename the columns in the reduced data set to names that are appropriate descriptors of the information contained in each variable according to the codebook. You may find the `.rename` method to be useful. Use the following variable names: `country`, `life exp`, `infant mortality`, `phys density`, `hospital bed`, `health exp` `percent GDP`, `OOP percent exp`, `health exp PC`, `fertility rate`, `GNI PC`, `regionname`.

### <font color = '#FF8C00'> Finding the Shape of the Dataframe </font>
We start off by finding the shape of the data frame, which means we get information as to how many rows and columns are in the dataset. Using `.shape` returns a tuple, where the first element is the number of rows and the second element is the number of columns.

In [ ]:
# printing out the shape of the data frame and the rows and columns
print("The Shape of the Dataframe: ", "\n")
print("Number of Rows: ", who_df.shape[0], "\n")
print("Number of Columns: ", who_df.shape[1])

The Shape of the Dataframe:  

Number of Rows:  193 

Number of Columns:  267


### <font color = '#FF8C00'> The Codebook & Dropping the Data Columns </font>
It can be seen in the dataset that there are large number of columns. We need a small number of them, so we drop the columns that are not mentioned in the codebook and keep the ones that are.

In [12]:
# the list of columns in the data set that are included in the code book
col_list = ["country", "v9", "v22", "v159", "v168", "v174", "v186", "v192", "v249", "v259", "regionname"]

# we take these columns of interest and create a new data frame
new_who_df = who_df[col_list]

# displaying the new data frame with the first 5 columns
new_who_df.head(5)


country  v9  v22  v159  v168  v174   v186    v192  v249    v259  \
0  Afghanistan  42  165   2.0   4.0   9.2   78.5    91.0   7.1     NaN   
1      Albania  72   13  12.0  30.0   6.5   94.9   381.0   2.1  6580.0   
2      Algeria  71   33  11.0  17.0   4.2   94.6   315.0   2.4  7640.0   
3      Andorra  81    3  36.0  32.0   7.4   73.4  2980.0   1.3     NaN   
4       Angola  53  116   1.0   8.0   2.6  100.0   115.0   6.5  4400.0   

  regionname  
0       Asia  
1     Europe  
2     Africa  
3     Europe  
4     Africa

### <font color = '#FF8C00'> Renaming the Data Columns </font>
We can now work on renaming the data columns. We rename the data columns as follows:
- `country` as `country`
- `v9` as `life_exp`
- `v22` as `infant_mortality`
- `v159` as `phys_density`
- `v168` as `hospital_bed`
- `v174` as `health_exp_percent_GDP`
- `v186` as `OOP_percent_exp`
- `v192` as `health_exp_PC`
- `v249` as `fertility_rate`
- `v259` as `GNI_PC`
- `regionname` as `regionname`


In [ ]:
# renaming the columns as mentioned in the code book
new_who_df = new_who_df.rename(columns={'v9' : 'life_exp'})

# displaying the new data frame with renamed columns
new_who_df

### <font color='#FF8C00'> Sources Used For Section One </font>
- https://stackoverflow.com/questions/16616141/keep-certain-columns-in-a-pandas-dataframe-deleting-everything-else
- https://stackoverflow.com/questions/11346283/renaming-column-names-in-pandas